app/code/Magento/Indexer/Model/Processor.php

```diff
@@ -60,32 +65,83 @@ public function __construct(
             /** @var Indexer $indexer */
             $indexer = $this->indexerFactory->create();
             $indexer->load($indexerId);
             $indexerConfig = $this->config->getIndexer($indexerId);
+            $sharedIndex = $indexerConfig['shared_index'];
+
             if ($indexer->isInvalid()) {
                 // Skip indexers having shared index that was already complete
                 $sharedIndex = $indexerConfig['shared_index'] ?? null;
-                if (!in_array($sharedIndex, $sharedIndexesComplete)) {
+                if (!in_array($sharedIndex, $this->sharedIndexesComplete)) {
                     $indexer->reindexAll();
-                } else {
-                    /** @var \Magento\Indexer\Model\Indexer\State $state */
-                    $state = $indexer->getState();
-                    $state->setStatus(StateInterface::STATUS_WORKING);
-                    $state->save();
-                    $state->setStatus(StateInterface::STATUS_VALID);
-                    $state->save();
-                }
-                if ($sharedIndex) {
-                    $sharedIndexesComplete[] = $sharedIndex;
+
+                    if ($sharedIndex) {
+                        $this->validateSharedIndex($sharedIndex);
+                    }
                 }
             }
         }
     }
 
+    /**
+     * Get indexer ids that have common shared index
+     *
+     * @param string $sharedIndex
+     * @return array
+     */
+    private function getIndexerIdsBySharedIndex(string $sharedIndex): array
+    {
+        $indexers = $this->config->getIndexers();
+
+        $result = [];
+        foreach ($indexers as $indexerConfig) {
+            if ($indexerConfig['shared_index'] == $sharedIndex) {
+                $result[] = $indexerConfig['indexer_id'];
+            }
+        }
+
+        return $result;
+    }
+
+    /**
+     * Validate indexers by shared index ID
+     *
+     * @param string $sharedIndex
+     * @return $this
+     */
+    private function validateSharedIndex(string $sharedIndex): self
+    {
+        if (empty($sharedIndex)) {
+            throw new \InvalidArgumentException(
+                'The sharedIndex is an invalid shared index identifier. Verify the identifier and try again.'
+            );
+        }
+
+        $indexerIds = $this->getIndexerIdsBySharedIndex($sharedIndex);
+        if (empty($indexerIds)) {
+            return $this;
+        }
+
+        foreach ($indexerIds as $indexerId) {
+            /** @var \Magento\Indexer\Model\Indexer $indexer */
+            $indexer = $this->indexerFactory->create();
+            $indexer->load($indexerId);
+            /** @var \Magento\Indexer\Model\Indexer\State $state */
+            $state = $indexer->getState();
+            $state->setStatus(StateInterface::STATUS_WORKING);
+            $state->save();
+            $state->setStatus(StateInterface::STATUS_VALID);
+            $state->save();
```

Comment:
Hi, why do we need to set status twice and save? we are currently having issue with indexes in 235 and came to this PR but i did not understand why this is set twice here.